crates/query/src/lib.rs

```diff
@@ -0,0 +1,265 @@
+//! Query scheduler and async runtime.
+
+pub mod arena;
+mod just_about_anything;
+pub mod name;
+
+use std::{
+    any::type_name,
+    fmt::Debug,
+    future::Future,
+    hash::{BuildHasherDefault, Hash},
+    pin::Pin,
+    sync::OnceLock,
+    task::{Context, Poll, RawWaker, RawWakerVTable, Waker},
+};
+
+use arena::{Arena, OwnPinned};
+use dashmap::{DashMap, DashSet};
+use just_about_anything::JustAboutAnything;
+use name::Name;
+use parking_lot::Mutex;
+use rustc_hash::FxHasher;
+
+struct Cache<'a, Q>
+where
+    Q: Query,
+{
+    cells: DashMap<Q, &'a OnceLock<Q::Result>, BuildHasherDefault<FxHasher>>,
+    enqueued: DashSet<Q, BuildHasherDefault<FxHasher>>,
+}
+
+impl<'a, Q> Cache<'a, Q>
+where
+    Q: Query,
+{
+    fn new() -> Self {
+        Self {
+            cells: DashMap::default(),
+            enqueued: DashSet::default(),
+        }
+    }
+
+    fn cell(&self, arena: &'a Arena, computation: Q) -> &'a OnceLock<Q::Result> {
+        *self
+            .cells
+            .entry(computation.clone())
+            .or_insert_with(|| arena.alloc(OnceLock::default()))
+    }
+}
+
+/// Queues up tasks that are to be done by the compilation loop.
+pub struct Scheduler<'a> {
+    /// The scheduler's arena. It is used for allocating things for the scheduler's own use, as well
+    /// as letting queries allocate their own data (such as results).
+    pub arena: &'a Arena,
+    caches_by_type: DashMap<Name, &'a dyn JustAboutAnything<'a>, BuildHasherDefault<FxHasher>>,
+    erased_queue: Mutex<Vec<Box<dyn ErasedQuery>>>,
+
+    #[cfg(debug_assertions)]
+    compute_type_names: DashMap<Name, &'static str, BuildHasherDefault<FxHasher>>,
+}
+
+impl<'a> Scheduler<'a> {
+    /// Create a new scheduler.
+    pub fn new(arena: &'a Arena) -> Self {
+        Self {
+            arena,
+            caches_by_type: DashMap::default(),
+            erased_queue: Mutex::new(vec![]),
+
+            #[cfg(debug_assertions)]
+            compute_type_names: DashMap::default(),
+        }
+    }
+
+    fn cache<Q>(&self) -> &Cache<'a, Q>
+    where
+        Q: Query,
+    {
+        #[cfg(debug_assertions)]
+        {
+            if let Some(previous_type_name) =
+                self.compute_type_names.insert(Q::NAME, type_name::<Q>())
+            {
+                assert_eq!(
+                    type_name::<Q>(),
+                    previous_type_name,
+                    "hash collision occurred; try using a different NAME for one of these types"
```

Comment:
Maybe instead of write "for one of these types" just write what name is invalid? Because you know which name was a problem.

Reply:
`assert_eq` prints out the left and right-hand side of the `==` in its assertion message:
```
thread 'main' panicked at src/main.rs:2:5:
assertion `left == right` failed: hash collision occurred; try using a different NAME for one of these types
  left: "some::Type"
 right: "some::other::Type"
```